### ISSUE TITLE
A Tuple is a Value

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#692]
Functional programming is all about _values_ and transformation of values using _functions_.

Compared to other types Tuples have still a special role in Javaslang. To reflect my statement above in a consistent way, Tuple _has to be_ a [Value](https://github.com/javaslang/javaslang/blob/master/src/main/java/javaslang/Value.java) in Javaslang.

This is the current type hierarchy:

![javaslang-2 0 0](https://cloud.githubusercontent.com/assets/743833/10410566/0f472f6a-6f46-11e5-906d-5f480ea702c0.png)

Up to now I emphasized the viewpoint that _Tuples are Tuples_, not more and not less. Especially no other types should be derived from tuples, i.e. they should not be extended.

But what is the value of a Tuple? In my opinion, the value of a Tuple is the Tuple itself. This is different to all other `Value` types of Javaslang, which are always _wrappers_ of values in some way (e.g. Option, List, ...).

**Now the big question: Is there any benefit to see a Tuple as a Value of itself, i.e. a Value of a Tuple?**

One benefit is, that Tuple would inherit [all operations on Values](https://github.com/javaslang/javaslang/blob/master/src/main/java/javaslang/Value.java).

Is a Tuple not lightweight enough any more if it inherits from Value?

```java

Value<Tuple1<Integer>> val = Tuple.of(1);

class Tuple1<T> implements Tuple, Value<T> {
    ...
}
```

**Comment by ggalmazor** [comment]
When I am working with Tuples I often want to handle their contents the same way I would with a List: map a Tuple3 into a Tuple2, fold all objects into one...

I see Tuples as a sequence of objects that are heterogeneously typed. According to the current type hierarchy, imo they should extend Value.

I guess that they couldn't extend Traversable because a Tuple doesn't store objects with the same type, but in an imaginary Java where this is possible, I'd see Tuple at the same level as Seq.

**Comment by danieldietrich** [comment]
Thx, it is good to hear your view on that, I felt the same.

**Comment by danieldietrich** [comment]
There will be some changes...

1. Value.filter will return Option instead of throwing NoSuchElementException for some types, e.g. for Lazy, Either, ... and of cause Tuple. In general we should not rely on throwing exceptions. Instead we should provide an algebraic value in order to _not break the control flow_!
2. In order to "map a Tuple3 into a Tuple2", as @ggalmazor suggested, we will rework our Tuple*.map method. E.g. for Tuple2 it will look like this: `<U extends Tuple<U>> U map(Function2<? super T1, ? super T2, U> f)`
3. The base interface `Tuple` needs a generic self-type in order to implement Value for each of Tuple* individually: `interface Tuple<$ extends Tuple<$>> extends Value<$>`

**Comment by danieldietrich** [comment]
To resolve ambiguities, we will rename the `map` method mentioned above (in 2.) to `transform()`:

```java
// defined in class TupleN, for N = 0, ..., 8
<U extends Tuple<U>> U transform(FunctionN<? super T1, ? super T2, ..., ? super TN, U> f)
```

**Comment by danieldietrich** [comment]
The fact may lead to ambiguities, that (in my current prototype) `Tuple<U extends Tuple<U>>` is also of type `Iterable<U>`.

For example, the compiler can't decide which of the following Assertj methods it should use:

* `assertThat(T)`
* `assertThat(Interable<T>)`

**Comment by danieldietrich** [comment]
After creating a prototype implementation and fiddling around one day, this does not feel right **from the API perspective**. Too many stumbling blocks.

I will revert my changes and implement the _lessons learned_:

* Implementations of Value.filter(Predicate) should return Option instead of throwing NoSuchElementException
* Tuple needs a method `<U extends Tuple<U>> U transform(Function<? super TupleN, U> f)`

**Comment by danieldietrich** [comment]
The new transform API allows us to write

```java
Tuple2<Integer, String> tuple = Tuple.of(1, "2");
Tuple3<Integer, Double, String> transform = tuple.transform(t -> Tuple.of(t._1, 3.1415, t._2));
```

Given `transform()` we have all the freedom we need to transform tuples. It would be great to have a general, abstract view on tuples as sequence of heterogenous values. But currently Java does not allow us to express this (in a practical way).

More specifically `transform()` is the general form of all (transforming) sequential operations, like `split`, `take` (aka `limit`), `drop` (aka `split`) et al.

(Accessors like fold, reduce et al. are another story.)

IMO it does not make much sense to implement sequential operations for tuples because the domain is finite, unless to sequences. More concrete, wlog, an operation like `flatten` on a tuple of tuples is not possible because the domain is finite:

```
// Meh! Tuple9 does not exist!
((1, 2, 3, 4), (5, 6, 7, 8), (9)).flatten() = (1, 2, 3, 4, 5, 6, 7, 8, 9)
```

/cc @lukaseder

**Comment by lukaseder** [comment]
I see, you're being lured into declaring recursive generics again :)

> `Tuple<U extends Tuple<U>>`

I just have a gut feeling that you're going to regret this, as this will again make the `Tuple` type unusable outside of the Javaslang API itself [(see again this post)](http://blog.jooq.org/2013/06/28/the-dangers-of-correlating-subtype-polymorphism-with-generic-polymorphism/). Why does `Tuple` need to know its own subtype?

Also, why the hassle with `map()` vs. `transform()`? Both methods return a tuple. What if I simply want to...

```java
String string = tuple2.map((v1, v2) -> "A string");
```

i.e.

```java
public final <R> R map(Function2<T1, T2, R> function) { ... }
```

In other words, I'm wondering if you're about to add overly "complex" API that doesn't pull the weight of the use-case (yet).

**Comment by lukaseder** [comment]
(oh, I just saw https://github.com/javaslang/javaslang/pull/696)

**Comment by danieldietrich** [comment]
Thanks for the feedback. You're right, it was not a good idea declaring recursive generics.

> What if I simply want to...

Yes, I recognized it, too. However, both of `map` and `transform` are needed.

* `map` does map the components while preserving the 'container' `Tuple*`
* `transform` may change the result type

I think now all is fine. Thx for the review!!

**Comment by lukaseder** [comment]
Ah, I see. I guess I got `map()` wrong a couple of times. The "container" should be preserved indeed. Although, it's still a bit weird for a `map()` function to require to produce the "container" instead of just the "contained value". E.g.

```java
interface Stream<T> {
    <U> Stream<U> map(Function<? super T, ? extends U> function);
}
```

The `map()` function doesn't require the implementation to produce the container (such as `flatMap()`), but to produce the contained value. For e.g. `Tuple2`, I would then expect:

```java
interface Tuple2<T1, T2> {
    <U1, U2> Tuple2<U1, U2> map(
        Function<? super T1, ? extends U1> function,
        Function<? super T2, ? extends U2> function
    );
}
```

That feels more like the "ordinary" `map()`. What do you think?

**Comment by danieldietrich** [comment]
I like images.

```java
interface M<T> {
    <U> M<U> map(Function<? super T, ? extends U> f)
}
```

can be seen like this:

```
   M<T> -- map --> M<U>
     |               ^
     v               |
     T ----- f ----> U
```

Description of the arrows:

* `map()` is a function of some `M<T>`. It uses an `f` to produce a `M<U>`
* `map()` preserves the container type `M`
* the _mapper_ `f` maps a value of type `T` to a value of type `U`, where `T` and `U` are generic

The semantics is (spoken in code, without taking inheritance into account):

```
interface M<T> {

    <U> M<U> cons(U u);

    T get();

    default <U> M<U> map(Function<? super T, ? extends U> f) {
        return cons(f.apply(get()));
    }
}
```

**Comment by danieldietrich** [comment]
Correct, your example showed exactly this!

```
interface Tuple2<T1, T2> {
    <U1, U2> Tuple2<U1, U2> map(
        Function<? super T1, ? extends U1> function,
        Function<? super T2, ? extends U2> function
    );
}
```

The general form is

```java
<U1, ..., UN> M<U1, ..., UN> map(
    FunctionN<? super T1, ..., ? super TN, ? extends TupleN<U1, ..., UN> f);
```

(In our example it is a coincidence that `M` = `Tuple2` = return type of `f`.)

Or component-wise, like in your example:

```java
<U1, ..., UN> M<U1, ..., UN> map(
    Function<? super T1, ? extends U1> f1,
    ...,
    Function<? super TN, ? extends UN> fn);
```

**Comment by lukaseder** [comment]
OK, perhaps I was missing something, somewhere when you commented:

```java
<U extends Tuple<U>> U map(Function2<? super T1, ? super T2, U> f)
```

**Comment by danieldietrich** [comment]
Yes, you are right, let me explain:

> But what is the value of a Tuple? In my opinion, the value of a Tuple is the Tuple itself. This is different to all other Value types of Javaslang, which are always wrappers of values in some way (e.g. Option, List, ...).

Tuple is very special. I wanted it to extends my abstraction for values `interface Value<T>`. The point is that `Value<T>` has only one generic type parameter.

This implies for example that `class Tuple2<T1, T2> implements Value<Tuple2<T1, T2>>`. It was a technical thing, I wanted to have _one abstraction_ for values in general - which we actually have in Javaslang!

Examples:

* `interface Map<T, U> extends Value<Entry<T, U>>` (indeed it is `Traversable<Entry<T, U>>`)
* Either's `RightProjection<L, R>` is a `Value<R>`, Either's `LeftProjection<L, R>` is a `Value<L>`
* and so on...

Does this sound reasonable?

**Comment by danieldietrich** [comment]
At the long end, this led to the recursive declaration `Tuple<U extends Tuple<U>>` to have common methods in the base interface Tuple and to get the return types right.

And that led to ambiguities which are a no-go for a user-friendly API.

**Comment by danieldietrich** [comment]
Oh, cool! Here we have an article :-)

**Comment by danieldietrich** [comment]
The `<U extends Tuple<U>>` reflects the recursive nature I mentioned above...

**Comment by lukaseder** [comment]
Oh, I see - I suspect I confused your original intent (what you just explained, and what's in the title), and some comments in between (about tuple being a container like `Seq`). Sorry about my confusion

**Comment by danieldietrich** [comment]
I was confused by myself. Then I delete code and all is fine again :-)

**Comment by danieldietrich** [comment]
At least some of the Value interdace methods should be supported by Tuple.

Either we split the Value interface into separated interfaces (Conversible, Monadic, ....) or we find a way to let Tuple implement Value (better solution).

Tuple could be a Value by

* taking only the first component into account (Tuple0 = empty)
* using a recursive type approach
* letting TupleN classes implement Value instead of letting the Tuple interface extending Value
* ...

**Comment by danieldietrich** [comment]
**Tuples as collections of elements**

When treating a Tuple as collection of elements, it is helpful if the component types are homogenous. See https://hackage.haskell.org/package/tuples-homogenous-h98-0.1.1.0/docs/Data-Tuple-Homogenous.html.

If the components are not homogenous, the best we can do is to treat them as Objects.

**Tuples as composite objects**

This viewpoint is already described above - the value of a tuple is the tuple itself.

**Comment by danieldietrich** [comment]
The biggest problem of implementing **Tuples as collections of elements** is ... the `map` method.

Starting with `interface Tuple extends Value<Object>` leads to

```java
// Tuple0: Error / incompatible return type
<U> Tuple0 map(Function<? super Object, ? extends U> mapper);

// Tuple1: ok!
<U> Tuple1<U> map(Function<? super Object, ? extends U> mapper);

// Tuple2: Error / incompatible return type
<U> Tuple2<U, U> map(Function<? super Object, ? extends U> mapper);
```

To address the 'incompatible return type'  issue, we may introduce a generic bound type parameter `B`:

```java
interface Tuple<B> extends Value<B> {
}
```

This makes it even worse:

```java
// Tuple0 extends Value<Object>: ok, but instead of Object it should be any type
<U> Tuple0 map(Function<? super Object, ? extends U> mapper);

// Tuple1<T1> extends Value<T1>: ok!
<U> Tuple1<U> map(Function<? super Object, ? extends U> mapper);

// Tuple2<T1, T2> extends Value<Object> (should be lower bound of T1, T2):
// Error / incompatible return type
<U> Tuple2<U, U> map(Function<? super Object, ? extends U> mapper);
```

When we change the return type of `map(Function)` from `TupleN` to `Value<U>`, the only remaining problem with `interface Tuple extends Value<Object>` is an ambiguity regarding `Tuple1.map(Function)` and `Tuple1.map(Function1)`:

```java
<U1> Tuple1<U1> map(Function1<? super T1, ? extends U1> f);

<U> Value<U> map(Function<? super Object, ? extends U> mapper)

// ambiguous!
map(o -> o)
```

**Conclusion:**

* We need to provide a type hint `interface Tuple<T> extends Value<T>`, in order to fix the Tuple1 map issue
* With a type hint, we have a problem to let return map a homogenous `TupleN<U, ..., U>`

No solution here for that, yet...

**Comment by danieldietrich** [comment]
This would (probably) work:

* `Tuple gets an additional type parameter `interface Tuple<B> extends Value<B>`
* `Tuple0` could implement `Tuple` (unchecked)
* `Tuple1<T1>` could implement `Tuple<T1>`
* `TupleN` could implement `Tuple` (unchecked)

Cons:

* Tuple has additional type parameter
* Only Tuple1 can make use of this type parameter, Tuple0, Tuple2, ... need to implement unchecked methods without generics

Conclusion:

The cons mentioned above sound like an evil hack. Instead, we should implement only some of the `Value` interfaces and exclude `FilterMonadic`. In particular we implement:

* `javaslang.Iterable`
* `Convertible`
* `Printable`

These interfaces should be internal and just a mechanism to organize things a bit and share common interfaces between tuples and values.

**Comment by danieldietrich** [comment]
Should 'non-monadic' Values have a name?

E.g.

* javaslang.Object
* SemiValue
* HalfValue
* Non / NonValue / NonMonadicValue
* AntiValue
* VirtualValue
* CompositeValue
* AbstractValue
* ...

**Comment by ruslansennov** [comment]
> Maybe it is sufficient not to abstract over this special case and just let Tuple implement the interfaces Convertable, Printable, etc.

fine :)

**Comment by danieldietrich** [comment]
:+1:

**Comment by danieldietrich** [comment]
I had an Idea - it is:

* rename `javaslang.Iterable` to `javaslang.Immutable`.
* let `Immutable` additionally extend `Convertible` and `Printable`
* let `Tuple` extends `Immutable`
* let `Value` extends `Immutable` and `FilterMonadic`

Additionally move `sliding` et. al. from `Iterable` to `TraversableOnce`. These operations should not be part of each `Value`.

**Comment by danieldietrich** [comment]
A prototypical implementation showed, that it is not a good idea to let Tuple implement Iterable. It led to ambiguities when throwing a tuple into one of two methods:

```java
void m(T);
void m(Iterable<T>);
```

Tuple now has toSeq() which is sufficient for now.
